src/shared/particle_dynamics/solid_dynamics/constraint_dynamics.h

```diff
@@ -59,46 +59,47 @@ typedef DataDelegateInner<SolidParticles> SolidDataInner;
  * 			TODO: to clarify the treatment of particle position,
  * 			how to achieve consistency between velocity and position constraints.
  */
-template <class DynamicsIdentifier>
-class BaseMotionConstraint : public BaseLocalDynamics<DynamicsIdentifier>, public SolidDataSimple
+template <class DynamicsIdentifier, class DataDelegationType>
+class BaseMotionConstraint : public BaseLocalDynamics<DynamicsIdentifier>, public DataDelegationType
 {
-  public:
-    explicit BaseMotionConstraint(DynamicsIdentifier &identifier)
-        : BaseLocalDynamics<DynamicsIdentifier>(identifier), SolidDataSimple(identifier.getSPHBody()),
-          pos_(particles_->pos_), pos0_(particles_->pos0_),
-          n_(particles_->n_), n0_(particles_->n0_),
-          vel_(particles_->vel_), force_(particles_->force_), mass_(particles_->mass_){};
-
-    virtual ~BaseMotionConstraint(){};
-
-  protected:
-    StdLargeVec<Vecd> &pos_, &pos0_;
-    StdLargeVec<Vecd> &n_, &n0_;
-    StdLargeVec<Vecd> &vel_, &force_;
-    StdLargeVec<Real> &mass_;
+public:
+    explicit BaseMotionConstraint(DynamicsIdentifier& identifier)
+        : BaseLocalDynamics<DynamicsIdentifier>(identifier), DataDelegationType(identifier.getSPHBody()),
+        pos_(this->particles_->pos_), pos0_(this->particles_->pos0_),
+        n_(this->particles_->n_), n0_(this->particles_->n0_),
+        vel_(this->particles_->vel_), force_(this->particles_->force_), mass_(this->particles_->mass_) {};
+
+    virtual ~BaseMotionConstraint() {};
+
+protected:
+    StdLargeVec<Vecd>& pos_, & pos0_;
+    StdLargeVec<Vecd>& n_, & n0_;
+    StdLargeVec<Vecd>& vel_, & force_;
+    StdLargeVec<Real>& mass_;
```

Comment:
In the present code, both "C++" style
Type* Ptr;
and  "C" style 
Type *Ptr;
are fine. 
While, when declare several variables together, we can follow "C" style, viz. 
Type *Ptr_1, *Ptr_2, *Ptr_3;
for clarity. 
Other wise,  "C++" style 
Type* Ptr_1, * Ptr_2, * Ptr_3;
seems not so tidy.

Reply:
@Shuaihao-Zhang @ChiZhangatTUM SPHinXsys is using clang format by default.
[.clang-format](https://github.com/Xiangyu-Hu/SPHinXsys/blob/master/.clang-format)
you can setup your visual studio so that it follows the format when you saving files.

Reply:
Thanks. I will revise the format.

Reply:
> @Shuaihao-Zhang @ChiZhangatTUM SPHinXsys is using clang format by default. [.clang-format](https://github.com/Xiangyu-Hu/SPHinXsys/blob/master/.clang-format) you can setup your visual studio so that it follows the format when you saving files.

@Xiangyu-Hu You can forward this to all the developers as the format in new code is not consistent with the existing ones.

Reply:
I also learned. Thanks!

Reply:
@Xiangyu-Hu @ChiZhangatTUM Thanks!
I configured Clang format in my Visual Studio and formatted the code with it.